Releasing the revised Golang runtime

sciabarracom opened the pull request:
notable changes:
- removed the handshake and completely revised the health check
- changed the protocol, now logs are stdout and stderr, and output is in channel 3
- using a library for golang actions, PR submitted to the go client library
- using the "official" golang docker images as a basis
- solved synchronization issues with the output and error logs and the "log guards"

dubee commented:
@sciabarracom, thanks for all your hard work on this one. Will take a look.

csantanapr commented:
Great progress @sciabarracom 🎉

mrutkows commented:
@dubee @sciabarracom @csantanapr @houshengbo @daisy-ycguo If we intend to add this to our release, we should formalize that in our docs and have Vincent and Daisy review the build script and legal files (e.g., ASF headers/scancode, Notice file, etc.) and add it to our automated builds scripts under incubator-openwhisk-release repo.

vvraskin commented on misc/.gitignore:
```diff
@@ -0,0 +1,2 @@
+exec
+exec.json
```
Should we merge all nested .gitignore files into one file in the root directory?

vvraskin commented on openwhisk/_test/first/3/.gitignore:
```diff
@@ -0,0 +1 @@
+*
```
Do we have a chance to remove binary and zip from git and generate it on the flight when the test is run? (Did not have a chance to comment under them, that's why here)

sciabarracom replied:
those are required for the test - I need a BINARY in elf format to check if the binary is recognized...

vvraskin commented on openwhisk/actionProxy.go:
```diff
@@ -0,0 +1,97 @@
+	if theChannel != nil {
+		log.Println("terminating old action")
+		theChannel <- ""
+		theChannel = nil
```
Should the channel be closed instead of assigning it to nil? This will automatically unblock all listeners if any

sciabarracom replied:
actually I already did a BIG refactoring of the pipe_exec that is greatly simplifying those cases. It lives here https://github.com/sciabarracom/incubator-openwhisk-runtime-go/blob/new-executor/openwhisk/executor.go
I have not yet merged with the master. Currently the PR works with the cases listed in the test/test.t that covers all the use cases discussed with Rodric.

vvraskin replied:
Maybe it would make sense to merge the refactoring into this PR? That would make it easier to review it, to avoid commenting things that were already reworked...

sciabarracom replied:
Yes but I applied your comments to my refactoring. I have almost finished it.

sciabarracom commented on .gitignore:
```diff
@@ -0,0 +1,8 @@
+.vscode/
+vendor/
+hack
```
yup - those were my tests...

sciabarracom commented on openwhisk/actionProxy.go:
```diff
@@ -0,0 +1,97 @@
+	// cannot start, removing the action and retry
+	exeDir := fmt.Sprintf("./action/%d/", highestDir)
+	os.RemoveAll(exeDir)
+	reStartAction()
```
Yup. Those cases are worked on in my refactoring.

sciabarracom commented on openwhisk/actionProxy.go:
```diff
@@ -0,0 +1,97 @@
+	// terminate the logger
+	if theLogger != nil {
+		theLogger <- false
+		theLogger = nil
```
Same answer :)

sciabarracom commented:
Released the refactoring:
- reworked the "PipeExec" now called the Executor, consolidating all the channels and go routines in a single, easier to understand, class
- reworked the start/stop logic, now atomic (you start a new action THEN stop the old one)
- able to detect termination of the underlying action, then returning error so the system can cleanup the container
- replaced the docker base image with ibm/ubuntu:14.04

sciabarracom commented:
This other commit is in preparation for swift support - I am splitting in 2 images, a base one goproxy to be used as a base image for swift actions, and action-golang to be used for go actions. Since I am going to provide source support this one includes also the golang compiler.

rabbah commented on CREDITS.txt:
```diff
@@ -0,0 +1 @@
+Michele Sciabarra <michele@sciabarra.com>
```
FWIW I don't believe we'd one credit files like this anywhere else.
I think this is the only credit file I'm aware of although I haven't checked all other repos. https://github.com/apache/incubator-openwhisk/blob/master/CREDITS.txt

rabbah commented on core/go19action/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM ibmcom/ubuntu:14.04
```
any particular reason to use the ibm ubuntu image?

rabbah commented on core/goproxy/.gitignore:
```diff
@@ -0,0 +1 @@
+proxy
```
we tend to have one .gitignore file per repo at the root - would it work here?

rabbah commented on misc/HelloStress.jmx:
```diff
@@ -0,0 +1,156 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
note to self: is there a test added to run this and prevent it from rotting?
